include/drivers/gpio.h

```diff
@@ -226,13 +557,448 @@ static inline int gpio_pin_write(struct device *port, u32_t pin,
  * @param pin Pin number where data is read.
  * @param value Integer pointer to receive the data values from the pin.
  * @return 0 if successful, negative errno code on failure.
+ *
+ * @deprecated This function has been deprecated, please use gpio_pin_get,
+ * gpio_pin_get_raw instead.
  */
-static inline int gpio_pin_read(struct device *port, u32_t pin,
+__deprecated static inline int gpio_pin_read(struct device *port, u32_t pin,
 				u32_t *value)
 {
 	return gpio_read(port, GPIO_ACCESS_BY_PIN, pin, value);
 }
 
+/**
+ * @brief Get physical level of all input pins in a port.
+ *
+ * A low physical level on the pin will be interpreted as value 0. A high
+ * physical level will be interpreted as value 1. This function ignores
+ * GPIO_ACTIVE_LOW flag.
+ *
+ * Value of a pin with index n will be represented by bit n in the returned
+ * port value.
+ *
+ * @param port Pointer to the device structure for the driver instance.
+ * @param value Pointer to a variable where pin values will be stored.
+ *
+ * @retval 0 If successful.
+ * @retval -EIO I/O error when accessing an external GPIO chip.
+ */
+__syscall int gpio_port_get_raw(struct device *port, gpio_pins_t *value);
+
+static inline int z_impl_gpio_port_get_raw(struct device *port,
+					   gpio_pins_t *value)
+{
+	const struct gpio_driver_api *api =
+		(const struct gpio_driver_api *)port->driver_api;
+
+	return api->port_get_raw(port, value);
+}
+
+/**
+ * @brief Get logical level of all input pins in a port.
+ *
+ * Get logical level of an input pin taking into account GPIO_ACTIVE_LOW flag.
+ * If pin is configured as Active High, a low physical level will be interpreted
+ * as logical value 0. If pin is configured as Active Low, a low physical level
+ * will be interpreted as logical value 1.
+ *
+ * Value of a pin with index n will be represented by bit n in the returned
+ * port value.
+ *
+ * @param port Pointer to the device structure for the driver instance.
+ * @param value Pointer to a variable where pin values will be stored.
+ *
+ * @retval 0 If successful.
+ * @retval -EIO I/O error when accessing an external GPIO chip.
+ */
+static inline int gpio_port_get(struct device *port, gpio_pins_t *value)
+{
+	struct gpio_driver_data *const data = port->driver_data;
+	int ret;
+
+	ret = gpio_port_get_raw(port, value);
+	if (ret == 0) {
+		*value ^= data->invert;
```

Comment:
@andrewboie could you please double check if accessing driver data directly in the inline function run within the thread context doesn't violate requirements of userspace mode. If that's not OK I'll add an extra function to the driver API to get the value of data.invert, however that would mean longer execution time.